File: sections/14_core_patterns_mixed_contest/problems/q_event_attendance/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    cin >> n;
    vector<pair<int, int>> events(n);
    for (auto &[start, finish] : events) cin >> start >> finish;
    sort(events.begin(), events.end());
    priority_queue<int, vector<int>, greater<int>> finishes;
    int index = 0, answer = 0;
    long long day = 0;
    while (index < n || !finishes.empty()) {
        if (finishes.empty()) day = max(day, (long long)events[index].first);
        while (index < n && events[index].first <= day) {
            finishes.push(events[index].second);
            ++index;
        }
        while (!finishes.empty() && finishes.top() < day) finishes.pop();
        if (!finishes.empty()) {
            finishes.pop();
            ++answer;
            ++day;
        }
    }
    cout << answer << '\n';
}

File: sections/14_core_patterns_mixed_contest/problems/q_event_attendance/solution.py
import heapq
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    events = sorted(zip(data[1::2], data[2::2]))
    finishes = []
    index = answer = 0
    day = 0
    while index < len(events) or finishes:
        if not finishes:
            day = max(day, events[index][0])
        while index < len(events) and events[index][0] <= day:
            heapq.heappush(finishes, events[index][1])
            index += 1
        while finishes and finishes[0] < day:
            heapq.heappop(finishes)
        if finishes:
            heapq.heappop(finishes)
            answer += 1
            day += 1
    print(answer)


if __name__ == "__main__":
    main()
